providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/hook/DataCollectorHook.java

```diff
@@ -0,0 +1,168 @@
+package dev.openfeature.contrib.providers.gofeatureflag.hook;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import dev.openfeature.contrib.providers.gofeatureflag.GoFeatureFlagProvider;
+import dev.openfeature.contrib.providers.gofeatureflag.exception.InvalidEndpoint;
+import dev.openfeature.contrib.providers.gofeatureflag.exception.InvalidOptions;
+import dev.openfeature.contrib.providers.gofeatureflag.hook.events.Event;
+import dev.openfeature.contrib.providers.gofeatureflag.hook.events.Events;
+import dev.openfeature.contrib.providers.gofeatureflag.hook.events.EventsPublisher;
+import dev.openfeature.sdk.FlagEvaluationDetails;
+import dev.openfeature.sdk.Hook;
+import dev.openfeature.sdk.HookContext;
+import dev.openfeature.sdk.Reason;
+import dev.openfeature.sdk.exceptions.GeneralError;
+import lombok.extern.slf4j.Slf4j;
+import okhttp3.HttpUrl;
+import okhttp3.MediaType;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static java.net.HttpURLConnection.HTTP_BAD_REQUEST;
+import static java.net.HttpURLConnection.HTTP_OK;
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+
+/**
+ * DataCollectorHook is an OpenFeature Hook in charge of sending the usage of the flag to GO Feature Flag.
+ */
+@Slf4j
+public class DataCollectorHook implements Hook {
+    public static final long DEFAULT_FLUSH_INTERVAL_MS = Duration.ofMinutes(1).toMillis();
+    public static final int DEFAULT_MAX_PENDING_EVENTS = 10000;
+    /**
+     * options contains all the options of this hook.
+     */
+    private final DataCollectorHookOptions options;
+    /**
+     * eventsPublisher is the system collecting all the information to send to GO Feature Flag.
+     */
+    private final EventsPublisher<Event> eventsPublisher;
+
+    /**
+     * Constructor of the hook.
+     *
+     * @param options - Options to configure the hook
+     * @throws InvalidOptions - Thrown when there is a missing configuration.
+     */
+    public DataCollectorHook(DataCollectorHookOptions options) throws InvalidOptions {
+        if (options == null) {
+            throw new InvalidOptions("No options provided");
+        }
+        if (options.getParsedEndpoint() == null) {
+            throw new InvalidEndpoint("endpoint is a mandatory field when creating the hook");
+        }
+        if (options.getFlushIntervalMs() != null && options.getFlushIntervalMs() <= 0) {
+            throw new InvalidOptions("flushIntervalMs must be larger than 0");
+        }
+        if (options.getMaxPendingEvents() != null && options.getMaxPendingEvents() <= 0) {
+            throw new InvalidOptions("maxPendingEvents must be larger than 0");
+        }
```

Comment:
Minor remark - this might best fit the `DataCollectorHookOptions` itself. For example, override lombok builder's builder and perform a validation there (fail fast)

Reply:
Good catch, I was not aware of this pattern with Lombok but I love it 😉